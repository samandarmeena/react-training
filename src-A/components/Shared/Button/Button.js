import React from 'react';
import PropTypes from 'prop-types';
import styles from './Button.module.scss';

/**
 * Button component
 * It renders button
 * @param {string} buttonType
 * @param {string} buttonName
 * @param {string} buttonId
 * @param {string} buttonTitle
 * @param {string} buttonStyle
 * @param {string} buttonSize
 * @param {string} addOnClass
 * @param {string} buttonToolTip
 * @param {fn} clickEvent
 * @return Element
 */
const Button = props => {
  const {
    buttonType,
    buttonName,
    buttonId,
    buttonTitle,
    buttonStyle,
    buttonSize,
    addOnClass,
    buttonToolTip,
    clickEvent
  } = props;
  return (
    <button
      type={buttonType}
      name={buttonName}
      id={buttonId}
      className={`${styles.button} ${styles[buttonStyle]} ${styles[buttonSize]} ${addOnClass}`}
      onClick={() => clickEvent()}
      title={buttonToolTip}
    >
      {buttonTitle}
    </button>
  );
};

// proptype validations for props
Button.propTypes = {
  buttonType: PropTypes.string,
  buttonName: PropTypes.string.isRequired,
  buttonId: PropTypes.string.isRequired,
  buttonTitle: PropTypes.oneOfType([PropTypes.string, PropTypes.element])
    .isRequired,
  buttonStyle: PropTypes.string,
  buttonSize: PropTypes.string,
  addOnClass: PropTypes.string,
  buttonToolTip: PropTypes.string.isRequired,
  clickEvent: PropTypes.func
};

// default values for props
Button.defaultProps = {
  buttonType: 'button',
  buttonStyle: 'buttonPrimary',
  buttonSize: 'buttonMedium',
  addOnClass: ''
};

export default Button;
