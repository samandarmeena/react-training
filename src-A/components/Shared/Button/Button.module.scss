.button {
  border: none;
  cursor: pointer;
  margin-left: 3px;
  margin-right: 3px;
}
.buttonInfo {
  background-color: #138496;
  color: #fff;
}
.buttonInfo:hover {
  color: #fff;
  background-color: #007183;
  border-color: #007183;
}
.buttonPrimary {
  background-color: #0069d9;
  color: #fff;
}
.buttonPrimary:hover {
  background-color: #0052c2;
}
.buttonSuccess {
  background-color: #40c2a6;
  color: #fff;
}
.buttonWarning {
  background-color: #e0a800;
  color: #fff;
}
.buttonDanger {
  background-color: #c82333;
  color: #fff;
}
.buttonDanger:hover {
  background-color: #ab0616;
}
.buttonLarge {
  font-size: 17px;
  border-radius: 5px;
  padding: 13px;
}
.buttonMedium {
  font-size: 15px;
  border-radius: 5px;
  padding: 10px;
}
.buttonSmall {
  font-size: 12px;
  border-radius: 5px;
  padding: 5px;
}
