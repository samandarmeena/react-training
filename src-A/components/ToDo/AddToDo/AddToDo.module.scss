.addToDo {
  font-size: 16px;
}
.title {
  font-size: 25px;
  margin-bottom: 10px;
}
.message {
  font-size: 15px;
  font-size: 17px;
  color: #4d4d4d;
  margin-bottom: 15px;
}
.welcomeBox {
  background: #ccc;
  padding: 30px 15px 40px 15px;
  border: 1px solid #b5b5b5;
  text-align: center;
  border-radius: 5px;
}
