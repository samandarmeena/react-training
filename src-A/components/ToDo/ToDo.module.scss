.toDoContainer {
  width: 50%;
  margin: auto;
  .addToDo {
    width: 100%;
  }
  .toDoListContainer {
    margin-top: 15px;
    border: 1px solid #ccc;
    padding: 10px;
    border-radius: 5px;
    .filterButton {
      width: 100%;
    }
  }
}
