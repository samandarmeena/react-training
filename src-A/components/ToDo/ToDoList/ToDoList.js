import React from 'react';
import PropTypes from 'prop-types';
import ToDoActions from './ToDoActions';
import styles from './ToDoList.module.scss';

/**
 * ToDoList component
 * renders ToDoList container
 * @param {array} todoItems
 * @param {string} filterBy
 * @param {func} toggleFixed
 * @param {func} removeToDo
 * @return Element
 */
const ToDoList = (props) => {
  const { todoItems, filterBy, toggleFixed, removeToDo } = props;
  return (
    <ul className={styles.toDoList}>
      {todoItems.map((todoItem) => {
        return todoItem.status === filterBy ? (
          <li key={todoItem.id} className={styles.listItems}>
            <span className={styles.toDoTitle}>{todoItem.title}</span>
            <span className={styles.toDoActions}>
              <ToDoActions
                itemId={todoItem.id}
                itemStatus={todoItem.status}
                handleToggleFixed={toggleFixed}
                handleRemoveToDo={removeToDo}
              />
            </span>
          </li>
        ) : (
          ''
        );
      })}
    </ul>
  );
};

// proptype validations for props
ToDoList.propTypes = {
  todoItems: PropTypes.array.isRequired,
  filterBy: PropTypes.string.isRequired,
  toggleFixed: PropTypes.func.isRequired,
  removeToDo: PropTypes.func.isRequired
};

export default ToDoList;
