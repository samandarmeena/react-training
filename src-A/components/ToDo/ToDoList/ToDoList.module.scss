.toDoList {
  padding: 0;
  .listItems {
    display: flex;
    font-size: 15px;
    border: 1px solid #ccc;
    margin-bottom: 10px;
    padding: 5px 10px;
    border-radius: 5px;
    align-items: center;
    justify-content: space-between;
    .toDoTitle {
      font-size: 15px;
    }
    .toDoActions {
      float: right;
    }
  }
}
