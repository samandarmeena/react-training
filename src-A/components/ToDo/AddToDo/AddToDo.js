import React from 'react';
import PropTypes from 'prop-types';
import Input from '../../Shared/Input';
import Button from '../../Shared/Button';
import styles from './AddToDo.module.scss';

/**
 * AddToDo component
 * renders AddToDo container
 * @param {string} todoTitle
 * @param {func} handleChange
 * @param {func} onSave
 * @return Element
 */
const AddToDo = (props) => {
  const { todoTitle, handleChange, onSave } = props;
  return (
    <div className={styles.welcomeBox}>
      <div className={styles.title}>Welcome!</div>
      <div className={styles.message}>
        To get started, add some items to your list.
      </div>
      <div className={styles.addToDo}>
        <Input
          inputType="text"
          inputName="todoTitle"
          inputId="todoTitle"
          inputPlaceholder="Enter here"
          inputValue={todoTitle}
          changeEvent={handleChange}
        />
        <Button
          buttonName="saveToDo"
          buttonId="saveToDo"
          buttonTitle="Add New"
          clickEvent={onSave}
          buttonToolTip="Add ToDo"
          addOnClass="saveToDoBtn"
        />
      </div>
    </div>
  );
};

// proptype validations for props
AddToDo.propTypes = {
  todoTitle: PropTypes.string.isRequired,
  handleChange: PropTypes.func.isRequired,
  onSave: PropTypes.func.isRequired
};

export default AddToDo;
