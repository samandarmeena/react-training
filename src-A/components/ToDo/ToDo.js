import React, { Component } from 'react';
import AddToDo from './AddToDo';
import ToDoList from './ToDoList';
import styles from './ToDo.module.scss';
import Button from '../Shared/Button';

/**
 * ToDo component
 * renders ToDo container
 * @return Element
 */
class ToDo extends Component {
  constructor(props) {
    super(props);
    this.state = { todoTitle: '', todos: [], filterBy: 'New' };
  }

  /** handle title value  */
  handleChange = (event) => {
    this.setState({ todoTitle: event.target.value });
  };

  /** save new todo */
  saveToDo = () => {
    if (this.state.todoTitle.trim()) {
      const newToDo = {
        id: this.state.todos.length + 1,
        title: this.state.todoTitle,
        status: 'New'
      };
      this.setState(
        { todos: [...this.state.todos, newToDo], filterBy: 'New' },
        () => {
          this.setState({ todoTitle: '' });
        }
      );
    }
  };

  /** remove a todo */
  removeToDo = (removeId) => {
    const newToDo = this.state.todos.filter((item) => item.id !== removeId);
    this.setState({
      todos: newToDo
    });
  };

  /** mark as completed/new */
  toggleFixed = (markId) => {
    const newToDo = this.state.todos.map((item) => {
      return item.id === markId
        ? {
            id: item.id,
            title: item.title,
            status: item.status === 'New' ? 'Completed' : 'New'
          }
        : item;
    });
    this.setState({
      todos: newToDo
    });
  };

  /** toggle between new and completed */
  toggleFilter = () => {
    let filterBy = this.state.filterBy === 'New' ? 'Completed' : 'New';
    this.setState({ filterBy: filterBy });
  };

  render() {
    return (
      <div className={styles.toDoContainer}>
        <div className={styles.addToDo}>
          <AddToDo
            todoTitle={this.state.todoTitle}
            handleChange={this.handleChange}
            onSave={this.saveToDo}
          />
        </div>
        {this.state.todos.length ? (
          <div className={styles.toDoListContainer}>
            <div className={styles.filterButton}>
              <Button
                buttonName="newToDO"
                buttonId="newToDO"
                buttonSize="buttonSmall"
                buttonStyle={
                  this.state.filterBy === 'New'
                    ? 'buttonPrimary'
                    : 'buttonSuccess'
                }
                addOnClass="buttonFilterBy"
                buttonToolTip="Toggle status"
                buttonTitle={this.state.filterBy}
                clickEvent={this.toggleFilter}
              />
            </div>
            <ToDoList
              todoItems={this.state.todos}
              removeToDo={this.removeToDo}
              toggleFixed={this.toggleFixed}
              filterBy={this.state.filterBy}
            />
          </div>
        ) : (
          ''
        )}
      </div>
    );
  }
}

export default ToDo;
