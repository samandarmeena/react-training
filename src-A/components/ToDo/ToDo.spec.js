import React from 'react';
import { mount } from 'enzyme';
import ToDo from './ToDo';
import AddToDo from '../ToDo/AddToDo';
import ToDoList from '../ToDo/ToDoList';
import Button from '../Shared/Button';

describe('ToDo component', () => {
  let wrapper = '';
  let mockState = {};
  beforeAll(() => {
    wrapper = mount(<ToDo />);
    mockState = {
      todoTitle: '',
      todos: [
        { id: 1, title: 'Test todo title 1', status: 'New' },
        { id: 2, title: 'Test todo title 2', status: 'New' }
      ],
      filterBy: 'New'
    };
  });

  it('ToDo component rendered without an error', () => {
    expect(wrapper).toBeTruthy();
  });

  it('Check length of the children', () => {
    expect(wrapper).toHaveLength(1);
  });

  it('Check children component rendered successfully', () => {
    expect(wrapper.find(AddToDo).length).toBe(1);
    expect(wrapper.find(ToDoList).length).toBe(0);
  });

  it('Check children component rendered successfully', () => {
    wrapper.setState(mockState);
    expect(wrapper.find(AddToDo).length).toBe(1);
    expect(wrapper.find('button.buttonFilterBy').length).toBe(1);
    expect(wrapper.find(ToDoList).length).toBe(1);
  });

  it('should render 2 items in the list', () => {
    wrapper.setState(mockState);
    const toDoList = wrapper.find(ToDoList);
    expect(toDoList.find('ul').find('li').length).toBe(2);
  });

  it('should change the filterby flag', () => {
    wrapper.setState(mockState);
    wrapper.find('button.buttonFilterBy').simulate('click');
    expect(wrapper.find('button.buttonFilterBy').text()).toBe('Completed');
  });

  it('should not add new todo in the list', () => {
    wrapper.find('button.buttonFilterBy').simulate('click');
    const input = wrapper.find('input');
    input.simulate('change', { target: { value: '' } });
    wrapper
      .find(AddToDo)
      .find('.addToDo')
      .find('button.saveToDoBtn')
      .simulate('click');
    const toDoList = wrapper.find(ToDoList);
    expect(toDoList.find('ul').find('li').length).toBe(2);
  });

  it('should add new todo in the list', () => {
    const input = wrapper.find('input');
    input.simulate('change', { target: { value: 'Test todo title 3' } });
    wrapper
      .find(AddToDo)
      .find('.addToDo')
      .find('button.saveToDoBtn')
      .simulate('click');
    const toDoList = wrapper.find(ToDoList);
    expect(toDoList.find('ul').find('li').length).toBe(3);
  });

  it('should change status of the todo', () => {
    wrapper
      .find(ToDoList)
      .find('ul')
      .find('li')
      .first()
      .find('button.changeStatus')
      .simulate('click');
    expect(
      wrapper
        .find(ToDoList)
        .find('ul')
        .find('li').length
    ).toBe(2);
  });

  it('one todo should have status completed', () => {
    wrapper.setState({ filterBy: 'Completed' });
    expect(
      wrapper
        .find(ToDoList)
        .find('ul')
        .find('li').length
    ).toBe(1);
  });

  it('toggle filter', () => {
    wrapper
      .find('.toDoListContainer')
      .find('.filterButton')
      .find(Button)
      .props()
      .clickEvent();
    expect(wrapper.state('filterBy')).toBe('New');
  });

  it('mark as new again', () => {
    wrapper
      .find('.toDoListContainer')
      .find('.filterButton')
      .find(Button)
      .props()
      .clickEvent();
    wrapper
      .find(ToDoList)
      .find('ul')
      .find('li')
      .first()
      .find('button.changeStatus')
      .simulate('click');
    expect(
      wrapper
        .find(ToDoList)
        .find('ul')
        .find('li').length
    ).toBe(0);
  });

  it('should remove a todo from the list', () => {
    wrapper.setState({ filterBy: 'New' });
    wrapper
      .find(ToDoList)
      .find('ul')
      .find('li')
      .at(1)
      .find('button.removeToDo')
      .simulate('click');
    expect(
      wrapper
        .find(ToDoList)
        .find('ul')
        .find('li').length
    ).toBe(2);
  });

  it('Snapshot testing', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
