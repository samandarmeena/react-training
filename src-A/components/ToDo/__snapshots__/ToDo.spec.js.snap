// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDo component Snapshot testing 1`] = `
<ToDo>
  <div
    className="toDoContainer"
  >
    <div
      className="addToDo"
    >
      <AddToDo
        handleChange={[Function]}
        onSave={[Function]}
        todoTitle=""
      >
        <div
          className="welcomeBox"
        >
          <div
            className="title"
          >
            Welcome!
          </div>
          <div
            className="message"
          >
            To get started, add some items to your list.
          </div>
          <div
            className="addToDo"
          >
            <Input
              changeEvent={[Function]}
              inputId="todoTitle"
              inputName="todoTitle"
              inputPlaceholder="Enter here"
              inputType="text"
              inputValue=""
            >
              <input
                className="inputBox"
                id="todoTitle"
                name="todoTitle"
                onChange={[Function]}
                placeholder="Enter here"
                type="text"
                value=""
              />
            </Input>
            <Button
              addOnClass="saveToDoBtn"
              buttonId="saveToDo"
              buttonName="saveToDo"
              buttonSize="buttonMedium"
              buttonStyle="buttonPrimary"
              buttonTitle="Add New"
              buttonToolTip="Add ToDo"
              buttonType="button"
              clickEvent={[Function]}
            >
              <button
                className="button buttonPrimary buttonMedium saveToDoBtn"
                id="saveToDo"
                name="saveToDo"
                onClick={[Function]}
                title="Add ToDo"
                type="button"
              >
                Add New
              </button>
            </Button>
          </div>
        </div>
      </AddToDo>
    </div>
    <div
      className="toDoListContainer"
    >
      <div
        className="filterButton"
      >
        <Button
          addOnClass="buttonFilterBy"
          buttonId="newToDO"
          buttonName="newToDO"
          buttonSize="buttonSmall"
          buttonStyle="buttonPrimary"
          buttonTitle="New"
          buttonToolTip="Toggle status"
          buttonType="button"
          clickEvent={[Function]}
        >
          <button
            className="button buttonPrimary buttonSmall buttonFilterBy"
            id="newToDO"
            name="newToDO"
            onClick={[Function]}
            title="Toggle status"
            type="button"
          >
            New
          </button>
        </Button>
      </div>
      <ToDoList
        filterBy="New"
        removeToDo={[Function]}
        todoItems={
          Array [
            Object {
              "id": 1,
              "status": "New",
              "title": "Test todo title 1",
            },
            Object {
              "id": 3,
              "status": "New",
              "title": "Test todo title 3",
            },
          ]
        }
        toggleFixed={[Function]}
      >
        <ul
          className="toDoList"
        >
          <li
            className="listItems"
            key="1"
          >
            <span
              className="toDoTitle"
            >
              Test todo title 1
            </span>
            <span
              className="toDoActions"
            >
              <ToDoActions
                handleRemoveToDo={[Function]}
                handleToggleFixed={[Function]}
                itemId={1}
                itemStatus="New"
              >
                <Button
                  addOnClass="changeStatus"
                  buttonId="complete1"
                  buttonName="complete1"
                  buttonSize="buttonSmall"
                  buttonStyle="buttonSuccess"
                  buttonTitle={
                    <Image
                      alt="Check"
                      src="/assets/images/check.png"
                    />
                  }
                  buttonToolTip="Mark as Completed"
                  buttonType="button"
                  clickEvent={[Function]}
                  key="complete1"
                >
                  <button
                    className="button buttonSuccess buttonSmall changeStatus"
                    id="complete1"
                    name="complete1"
                    onClick={[Function]}
                    title="Mark as Completed"
                    type="button"
                  >
                    <Image
                      alt="Check"
                      src="/assets/images/check.png"
                    >
                      <img
                        alt="Check"
                        src="/assets/images/check.png"
                      />
                    </Image>
                  </button>
                </Button>
                <Button
                  addOnClass="removeToDo"
                  buttonId="remove1"
                  buttonName="remove1"
                  buttonSize="buttonSmall"
                  buttonStyle="buttonDanger"
                  buttonTitle={
                    <Image
                      alt="Remove"
                      src="/assets/images/trash.png"
                    />
                  }
                  buttonToolTip="Remove ToDo"
                  buttonType="button"
                  clickEvent={[Function]}
                  key="remove1"
                >
                  <button
                    className="button buttonDanger buttonSmall removeToDo"
                    id="remove1"
                    name="remove1"
                    onClick={[Function]}
                    title="Remove ToDo"
                    type="button"
                  >
                    <Image
                      alt="Remove"
                      src="/assets/images/trash.png"
                    >
                      <img
                        alt="Remove"
                        src="/assets/images/trash.png"
                      />
                    </Image>
                  </button>
                </Button>
              </ToDoActions>
            </span>
          </li>
          <li
            className="listItems"
            key="3"
          >
            <span
              className="toDoTitle"
            >
              Test todo title 3
            </span>
            <span
              className="toDoActions"
            >
              <ToDoActions
                handleRemoveToDo={[Function]}
                handleToggleFixed={[Function]}
                itemId={3}
                itemStatus="New"
              >
                <Button
                  addOnClass="changeStatus"
                  buttonId="complete3"
                  buttonName="complete3"
                  buttonSize="buttonSmall"
                  buttonStyle="buttonSuccess"
                  buttonTitle={
                    <Image
                      alt="Check"
                      src="/assets/images/check.png"
                    />
                  }
                  buttonToolTip="Mark as Completed"
                  buttonType="button"
                  clickEvent={[Function]}
                  key="complete3"
                >
                  <button
                    className="button buttonSuccess buttonSmall changeStatus"
                    id="complete3"
                    name="complete3"
                    onClick={[Function]}
                    title="Mark as Completed"
                    type="button"
                  >
                    <Image
                      alt="Check"
                      src="/assets/images/check.png"
                    >
                      <img
                        alt="Check"
                        src="/assets/images/check.png"
                      />
                    </Image>
                  </button>
                </Button>
                <Button
                  addOnClass="removeToDo"
                  buttonId="remove3"
                  buttonName="remove3"
                  buttonSize="buttonSmall"
                  buttonStyle="buttonDanger"
                  buttonTitle={
                    <Image
                      alt="Remove"
                      src="/assets/images/trash.png"
                    />
                  }
                  buttonToolTip="Remove ToDo"
                  buttonType="button"
                  clickEvent={[Function]}
                  key="remove3"
                >
                  <button
                    className="button buttonDanger buttonSmall removeToDo"
                    id="remove3"
                    name="remove3"
                    onClick={[Function]}
                    title="Remove ToDo"
                    type="button"
                  >
                    <Image
                      alt="Remove"
                      src="/assets/images/trash.png"
                    >
                      <img
                        alt="Remove"
                        src="/assets/images/trash.png"
                      />
                    </Image>
                  </button>
                </Button>
              </ToDoActions>
            </span>
          </li>
        </ul>
      </ToDoList>
    </div>
  </div>
</ToDo>
`;
